Assignment5.java

```diff
@@ -0,0 +1,198 @@
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.LinkedList;
+import java.util.LinkedHashSet;
+import java.util.Collections;
+
+public class Assignment5 {
+    /**
+     * Infers the alphabet of an unknown language from list of words of that language.
+     *
+     * @param words list of words in lexicographic order.
+     * @return characters in lexicographic order.
+     */
+    public static List<Character> inferAlphabet(List<String> words) {
+        if (words.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        Set<Character> set = new HashSet<>();
+
+        String s1 = words.get(0);
+        // Performed so that all characters will be in the returned list even if topologicalSort doesn't contain some.
+        addAllCharacters(s1, set);
+
+        Map<Character, Vertex> graph = new HashMap<>();
+
+        for (int i = 0; i < words.size() - 1; i++) {
+            s1 = words.get(i);
```

Comment:
Be careful with using `get(index)` on a "plain" `List` - someone could pass `LinkedList` as parameter, which would give you O(n^2) complexity to just traverse the list.

The best solution would be to use `Iterator` (or `ListIterator`, if its features would be useful). Quite a bit worse solution would be to require `ArrayList` as parameter to have guarantees about complexity of `get(index)`.